### ISSUE TITLE
Low time effiency when run cnn on mnist-traning only with CPU 

### ISSUE THREAD

**Issue by Viewer-HX** [huggingface/candle#2144]
I try to use the following command to train the CNN model with mnist-training example. It works well when training linear and mlp model, but when I config it to train the CNN model, it needs approximately 1hr30mins per epoch. I think it may be abnormal. 

`cargo run --example mnist-training --features="candle-datasets" cnn --epochs=5 --save=cnn.safetensors`

Can you please try it?

**Comment by LaurentMazare** [comment]
You may want to compile with `--release` to enable optimizations?

**Comment by Viewer-HX** [comment]
@LaurentMazare 
Thanks for the suggestion. It works.
With `--release`, now the training speed reaches around 8min per epoch. Improve a lot.
But I'm still confused, 
(1) why the release mode is so fast? Is the optimization very aggressive?
(2)It still is slower than pytorch without GPU (around 50s per epoch.) What is the reason?
Here the training only uses CPU, is the algorithm too time-consuming?

Execution command
`cargo run --example mnist-training --release --features="candle-datasets" cnn --epochs=5 --save=cnn.safetensors`

**Comment by Viewer-HX** [comment]
Hi, I noticed an even weirder scenario, when I scale up the training epochs to 100 epochs, the training significantly decreases the accuracy.
training with GPU
Commands:
`cargo run --example mnist-training --release --features="candle-datasets" --features="cuda" cnn --epochs=100 --save=../../Saved_Models/rust_cnn.safetensors`


Results:
<img width="418" alt="Screenshot 2024-04-29 at 5 54 01 PM" src="https://github.com/huggingface/candle/assets/22735592/419289d5-540f-433c-8391-75859e2db12d">
<img width="582" alt="Screenshot 2024-04-29 at 5 54 35 PM" src="https://github.com/huggingface/candle/assets/22735592/ec9141f9-f608-4876-8ecc-1266ec3c9708">

**Comment by ArisuYuki** [comment]
Recently, I manually implemented a CNN for training mnists. The network structure is the same as in the example, but with pure CPU inference, one epoch takes about 14 seconds, and with CUDA acceleration, the one epoch takes about 0.4 seconds. I am very sure that there are some issues with the code in the example, because when I use "cargo run -- example mnist training -- release  --features  cnn" for training, it takes almost a minute for one round with pure CPU, which is strange
